crates/red_knot/src/module.rs

```diff
@@ -0,0 +1,708 @@
+#![allow(unreachable_pub)]
+
+use std::fmt::Formatter;
+use std::path::{Path, PathBuf};
+use std::sync::atomic::AtomicU32;
+use std::sync::Arc;
+
+use crate::files::{FileId, Files};
+use crate::FxDashMap;
+use dashmap::mapref::entry::Entry;
+
+/// ID uniquely identifying a module.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub struct ModuleId(u32);
+
+/// A module name, e.g. `foo.bar`.
+///
+/// Always normalized to the absolute form (never a relative module name).
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub struct ModuleName(smol_str::SmolStr);
+
+#[allow(unused)]
+impl ModuleName {
+    pub fn new(name: &str) -> Self {
+        debug_assert!(!name.is_empty());
+
+        Self(smol_str::SmolStr::new(name))
+    }
+
+    pub fn relative(_dots: u32, name: &str, _to: &Path) -> Self {
+        // FIXME: Take `to` and `dots` into account.
+        Self(smol_str::SmolStr::new(name))
+    }
+
+    pub fn from_relative_path(path: &Path) -> Option<Self> {
+        let name = if let Some(parent) = path.parent() {
+            let mut name = String::new();
+
+            for component in parent.components() {
+                if !name.is_empty() {
+                    name.push('.');
+                }
+
+                name.push_str(component.as_os_str().to_str()?);
+            }
+
+            // SAFETY: Unwrap is safe here or `parent` would have returned `None`.
+            name.push_str(path.file_stem().unwrap().to_str()?);
```

Comment:
Nice find. I now pop `__init__` at the entry of the function. 

> (In Python I'd avoid the need to manually add/remove dots by building up a list of strings instead of a single string, and then joining them on dot once done. But that may not be quite as efficient.)

Yeah, you can do this in Rust too. I now added a small optimization to allocate the right size of path so that it now should be faster :P